File: scala_code/src/main/scala/com/artvb/iq/chapter1/problem4/UrlSpaceEncoder.scala
package com.artvb.iq.chapter1.problem4

/**
 * Created by artburkart on 10/27/14.
 */
object UrlSpaceEncoder {
  /** URL encodes spaces as "%20" and trims off the extra spaces on the end
    *
    * @param str given String to URL encode spaces in
    * @param trueLength given length of String to URL encode
    * @return returns String with spaces URL encoded
    */
  def urlSpaceEncoder(str: String, trueLength: Int): String = {
    def urlSpaceEncoder1(l: List[Char], acc: String = ""): String = l match {
      case (Nil) => acc
      case (h :: t) => if (h == ' ') urlSpaceEncoder1(t, acc + "%20") else urlSpaceEncoder1(t, acc + h)
    }
    urlSpaceEncoder1(str.toList take trueLength)
  }
}

File: java_code/src/main/java/com/artvb/iq/chapter1/problem4/UrlSpaceEncoder.java
package com.artvb.iq.chapter1.problem4;

/**
 * Created by artburkart on 10/27/14.
 */
public class UrlSpaceEncoder {
    /**
     * URL encodes spaces as "%20" and trims off the extra spaces on the end
     *
     * @param str given String to URL encode spaces in
     * @param trueLength given length of String to URL encode
     * @return returns String with spaces URL encoded
     */
    public static String urlSpaceEncoder(String str, int trueLength) {
        if (str == null) { throw new NullPointerException(); }
        if (trueLength < 0) { throw new IllegalArgumentException(); }

        if (str.length() == 0) {
            return str;
        }

        char[] chars = str.toCharArray();
        int placeholder = chars.length - 1;
        for (int i = trueLength - 1; i >= 0; i--) {
            if (placeholder < 0) {
                break;
            } else if (chars[i] == ' ') {
                chars[placeholder] = '0';
                chars[placeholder - 1] = '2';
                chars[placeholder - 2] = '%';
                placeholder -= 3;
            } else {
                chars[placeholder] = chars[i];
                placeholder -= 1;
            }
        }
        return new String(chars);
    }
}
